ledger/participant-integration-api/src/main/scala/platform/partymanagement/PersistentPartyRecordStore.scala

```diff
@@ -0,0 +1,301 @@
+              Future.successful(
+                Left(PartyRecordStore.PartyNotFound(party))
+              )
+            }
+        } yield createdPartyRecord
+    }
+  }
+
+  override def getPartyRecord(
+      party: Party
+  )(implicit loggingContext: LoggingContext): Future[Result[ParticipantParty.PartyRecord]] = {
+    inTransaction(_.getPartyRecord) { implicit connection =>
+      withPartyRecord(party) { dbPartyRecord =>
+        val annotations = backend.getPartyAnnotations(dbPartyRecord.internalId)(connection)
+        toDomainPartyRecord(dbPartyRecord.payload, annotations)
+      }
+    }
+  }
+
+  private def doCreatePartyRecord(
+      partyRecord: ParticipantParty.PartyRecord
+  )(connection: Connection): Unit = {
+    val now = epochMicroseconds()
+    val dbParty = PartyRecordStorageBackend.DbPartyRecordPayload(
+      party = partyRecord.party,
+      resourceVersion = 0,
+      createdAt = now,
+    )
+    val internalId = backend.createPartyRecord(
+      partyRecord = dbParty
+    )(connection)
+    partyRecord.metadata.annotations.foreach { case (key, value) =>
+      backend.addPartyAnnotation(
+        internalId = internalId,
+        key = key,
+        value = value,
+        updatedAt = now,
+      )(connection)
+    }
+  }
+
+  private def createAndUpdatePartyRecordInTransaction(
+      partyRecordUpdate: PartyRecordUpdate,
+      party: Party,
+      newPartyRecord: ParticipantParty.PartyRecord,
+  )(implicit loggingContext: LoggingContext): Future[Result[ParticipantParty.PartyRecord]] = {
+    inTransaction(_.createPartyRecordOnUpdate) { implicit connection =>
+      for {
+        _ <- withoutPartyRecord(id = newPartyRecord.party) {
+          doCreatePartyRecord(newPartyRecord)(connection)
+        }
+        _ <- withPartyRecord(party) { dbPartyRecord =>
+          doUpdatePartyRecord(
+            dbPartyRecord = dbPartyRecord,
+            partyRecordUpdate = partyRecordUpdate,
+          )(connection)
+        }
+        updatePartyRecord <- {
+          withPartyRecord(party) { updatedDbPartyRecord =>
+            val annotations =
+              backend.getPartyAnnotations(updatedDbPartyRecord.internalId)(connection)
+            toDomainPartyRecord(
+              payload = updatedDbPartyRecord.payload,
+              annotations = annotations,
+            )
+          }
+
+        }
+      } yield updatePartyRecord
+    }.map(tapSuccess { newPartyRecord =>
+      logger.error(
+        s"Party record to update didn't exist so created a new party record in participant local store: ${newPartyRecord}"
+      )
+    })(scala.concurrent.ExecutionContext.parasitic)
+  }
+
+  private def doUpdatePartyRecord(
+      dbPartyRecord: PartyRecordStorageBackend.DbPartyRecord,
+      partyRecordUpdate: PartyRecordUpdate,
+  )(connection: Connection): Unit = {
+    val now = epochMicroseconds()
+    // Step 1: Update resource version
+    // NOTE: We starts by writing to the 'resource_version' attribute
```

Comment:
For competing update operations will not observe not-repeatable reads or phantom-reads because the first query they all do is to try to update the resource version on the same row which effectively acts like a mutex.

For get-style operations: 1) reading a party record row and 2) reading all annotations for a party record are both single queries, so they individually each see a consistent view of the data.
An inconsistent view of the data can across those two queries. But that's inconsequential for party records, as party record table does not have modifiable fields. It however might be consequential for participant users where we have two modifiable attributes: primaryParty and isDeactivated.
But that is not a new problem in user management where we have had modifiable user rights from the outset.

Reply:
I agree with your analysis, and let's not worry about this at the moment. FYI @nmarton-da